snippets/csharp/VS_Snippets_CLR_System/system.collections.generic.equalitycomparer.default/cs/program.cs

```diff
@@ -1,128 +1,96 @@
 
-        Dictionary<Box, string> boxesByDim = new Dictionary<Box, string>(bxd);
-        Dictionary<Box, string> boxesByVol = new Dictionary<Box, string>(bxv);
+        var boxes = new CustomCollection<Box>(new[] { redBox, blueBox, greenBox });
 
-        try
-        {
-            Box redBox = new Box(8, 8, 4);
-            Box blueBox = new Box(6, 8, 4);
-            Box greenBox = new Box(4, 8, 8);
+        var foundByDimension = boxes.FindFirst(greenBox);
 
-            Console.WriteLine("Adding boxes by Dimension");
+        Console.WriteLine($"Found box {foundByDimension} by dimension.");
 
-            boxesByDim.Add(redBox, "red");
-            boxesByDim.Add(blueBox, "blue");
-            boxesByDim.Add(greenBox, "green");
+        var foundByVolume = boxes.FindFirst(greenBox, new BoxEqVolume());
 
-            PrintBoxCollection(boxesByDim);
+        Console.WriteLine($"Found box {foundByVolume} by volume.");
+    }
+}
 
-            Console.WriteLine("\nAdding boxes by Volume");
+public class CustomCollection<T>
+{
+    private IEnumerable<T> items;
 
-            boxesByVol.Add(redBox, "red");
-            boxesByVol.Add(blueBox, "blue");
-            boxesByVol.Add(greenBox, "green");
+    public CustomCollection(IEnumerable<T> items) => this.items = items;
 
-            PrintBoxCollection(boxesByVol);
-        }
-        catch (ArgumentException argEx)
-        {
-            Console.WriteLine(argEx.Message);
-        }
-    }
-    private static void PrintBoxCollection(Dictionary<Box,string> boxes)
+    public T FindFirst(T itemToFind, IEqualityComparer<T> comparer = null)
     {
-        foreach (KeyValuePair<Box, string> kvp in boxes)
+        comparer = comparer ?? EqualityComparer<T>.Default;
+
+        foreach (var item in items)
         {
-            Console.WriteLine("{0} x {1} x {2} - {3}", kvp.Key.Height.ToString(),
-                                                       kvp.Key.Length.ToString(),
-                                                       kvp.Key.Width.ToString(), kvp.Value);
+            if (comparer.Equals(item, itemToFind))
+            {
+                return item;
+            }
         }
-    }
-}
 
-public class BoxEqDimensions : EqualityComparer<Box>
-{
-    public override int GetHashCode(Box bx)
-    {
-        int hCode = bx.Height ^ bx.Length ^ bx.Width;
-        return hCode.GetHashCode();
-    }
-    
-    public override bool Equals(Box b1, Box b2)
-    {
-        return EqualityComparer<Box>.Default.Equals(b1, b2);
+        throw new InvalidOperationException("No macthing item found.");
     }
 }
 
-
 public class BoxEqVolume : EqualityComparer<Box>
 {
-    public override int GetHashCode(Box bx)
-    {
-        int hCode = bx.Height ^ bx.Length ^ bx.Width;
-        return hCode.GetHashCode();
-    }
-
     public override bool Equals(Box b1, Box b2)
     {
-        if (b1.Height * b1.Width * b1.Length ==
-            b2.Height * b2.Width * b2.Length)
-        {
+        if (b1 == b2)
```

Comment:
It's hard to reason about this method, because it's not clear what `==` does for boxes. What if `==` gets overridden (to match `IEquatable<Box>` behavior, for exampe)? Did you mean here:
```csharp
if (object.ReferenceEquals(b1, b2))
   return true;
```

Reply:
Done. I have also changed `== null` to `is null` for similar reason.